src/ethereum/frontier/state.py

```diff
@@ -333,3 +350,47 @@ def write_code(sender: Account) -> None:
         sender.code = code
 
     modify_state(state, address, write_code)
+
+
+def snapshot_state(state: State) -> State:
```

Comment:
It was my understanding that everything in the state was frozen already, so couldn't you just do a single level copy?

Reply:
Yeah, that's the first thing I tried. However a simple deepcopy on state results in an error:

Option 1: Use copy
```
>>> state_snapshot = copy(state)
>>> acc_2 = bytes.fromhex("095e7baea6a6c7c4c2dfeb977efac326af552d00")
>>> increment_nonce(state, acc_2)
>>> assert state_snapshot != state
AssertionError                            Traceback (most recent call last)
<ipython-input-20-698273bffc21> in <module>
----> 1 assert new_state != state

AssertionError:
```
in the above example changes made to `state` are also applied to `state_snapshot`.

Option 2: Use deepcopy
```python
>>> from copy import copy, deepcopy
>>>  from ethereum.frontier.state import State, set_account, EMPTY_ACCOUNT, increment_nonce
>>> state = State()
>>> acc_1 = bytes.fromhex("095e7baea6a6c7c4c2dfeb977efac326af552d87")
>>> set_account(state, acc_1, EMPTY_ACCOUNT)
>>> deepcopy(state)
~/Python-Envs/eth-spec/lib/python3.7/copy.py in _reconstruct(x, memo, func, args, state, listiter, dictiter, deepcopy)
    290             if slotstate is not None:
    291                 for key, value in slotstate.items():
--> 292                     setattr(y, key, value)
    293
    294     if listiter is not None:

~/Projects/eth1.0-specs/src/ethereum/base_types.py in _setattr_function(self, attr, value)
    696
    697 def _setattr_function(self: Any, attr: str, value: Any) -> None:
--> 698     if self._frozen:
    699         raise Exception("Mutating frozen dataclasses is not allowed.")
    700     else:

AttributeError: _frozen
```

Which is why I had to come up with a custom implementation of deepcopy.

Reply:
@ultratwo any thoughts on this?

Reply:
A two layer copy is needed. `copy()` is two shallow, but your `deepcopy()` unnecessarily goes a layer too far and copies all the frozen stuff.

I think the best way to proceed is to implement the following interface, with the obvious semantics.
```
start_transaction()
commit_transaction()
abort_transaction()
```
I'll go implement that now.

Reply:
Cool, it would be awesome to have a `commit` `rollback` type interface.  But that shouldn't block this PR from being merged.